### PR TITLE
NUMA Manager - Generic bit calculations

### PR DESCRIPTION
**Pull request by nolancon**
CPU Manager and Device Manager create 2D-slices of NUMA affinity for requested resources. NUMA Manger takes individual affinities and calculates overall NUMA affinity for pod using generic mechanism.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 1]
```diff
@@ -197,56 +198,77 @@ func (m *manager) GetNUMAHints(resource string, amount int) numamanager.NumaMask
...
+		if CPUsInSocketSize[i] >= amount64 {
+			for j := 0; j < i; j++ {
+				arr = append(arr, 0)
+			}
+			arr = append(arr, 1)
+			
+			for j := 0; j < socketCount-(i+1); j++ {
+				arr = append(arr, 0)
+			}
+		} else {
```
Comment (lmdaly) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go]: Do we need this else statement? Just adding 0 values.. I guess could be removed

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 2]
```diff
@@ -172,60 +172,95 @@ func (m *ManagerImpl) GetNUMAHints(resource string, amount int) numamanager.Numa
...
+    	glog.Infof("[device-manager] Available devices for resource: %v", available)	
+	duplicate_frequency := make(map[int64]int64)
+	var socketValues []int64	
+ 	for availID := range available {
+		for _, device := range devices[resource] {
+			glog.Infof("[device-manager] AvailID: %v DeviceID: %v", availID, device)
+			if availID == device.ID {
+				socket := int64(device.Socket)
+				socketValues = append(socketValues, socket)	//slice of sockets with resource
+				if duplicate_frequency[socket] <= 1 {
```
Comment (lmdaly) [inline pkg/kubelet/cm/devicemanager/manager.go]: Should this be >= 1? 
if duplicate_frequncy[socket] = nil, want to initalize it as 1 else if its anything greater than or equal than one we want to increment

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 3]
```diff
@@ -172,60 +172,95 @@ func (m *ManagerImpl) GetNUMAHints(resource string, amount int) numamanager.Numa
...
+                }
+        }
+        glog.Infof("[device-manager] Largest socket that is NOT empty: %v", largestSkt)
+	
+	var socketMask []int64
+	fullMask:= make([][]int64,len(availableSockets))
+	var socket int64 = 0
+	var i int64
+	// Use largest socket to construct 2D slice (mask) for return to NUMA Manager
+	for j := 0; j < len(availableSockets); j++ {
```
Comment (lmdaly) [inline pkg/kubelet/cm/devicemanager/manager.go]: You could do something like 
for j, socket := range available sockets {
    //calculation on socket
}

Means you get the socket assigned without explicitly doing so in the loop

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 4]
```diff
@@ -172,60 +172,95 @@ func (m *ManagerImpl) GetNUMAHints(resource string, amount int) numamanager.Numa
...
+        	}			
+		fullMask[j] = socketMask 
+	}	
+	
+	//Build final slice for full mask - overall affinity for preferred state (only if more than 1 available skt)
+	overallMask := make([]int64, largestSkt+1)
+	if (len(fullMask) > 1) || ((len(fullMask) == 0) && (devicesTotal >= amount64)) {
+		var k int64
+		for k = 0; k < largestSkt+1; k ++ {
+			for l := 0; l < len(notEmptySkts); l ++ {
```
Comment (lmdaly) [inline pkg/kubelet/cm/devicemanager/manager.go]: Can use that array here as well.
for _, nonEmptySocket := range notEmptySockets {
    if k == nonEmptySocket
}

File: pkg/kubelet/cm/numamanager/numamanger.go [hunk 5]
```diff
@@ -66,75 +68,151 @@ func (m *numaManager) GetAffinity(podUID string, containerName string) NumaMask
...
-                    }
-                }
-            } else if numaMask.Affinity && numaMask.Mask == nil {
-                glog.Infof("[numamanager] NO Numa Affinity.")
-                return NumaMask {
-                    Mask:		nil,
-                    Affinity:	true,
-                }
-            }
+        var row,col int = 1, 1
```
Comment (lmdaly) [inline pkg/kubelet/cm/numamanager/numamanger.go]: Do you need this?

Could have nil mask sent back?

File: pkg/kubelet/cm/numamanager/numamanger.go [hunk 6]
```diff
@@ -66,75 +68,151 @@ func (m *numaManager) GetAffinity(podUID string, containerName string) NumaMask
...
+        
+	var maskHolder []string
+	count := 0 
+	var finalMask int64
+        for _, hp := range m.hintProviders {
+		for resource, amount := range container.Resources.Requests {
+			glog.Infof("Container Resource Name in NUMA Manager: %v, Amount: %v", resource, amount.Value())
+			numaMask := hp.GetNUMAHints(string(resource), int(amount.Value()))                          
+			if numaMask.Affinity && numaMask.Mask != nil {
+				if count == 0 {
```
Comment (lmdaly) [inline pkg/kubelet/cm/numamanager/numamanger.go:84]: Maybe this could be a method too to calculate the MaskHolder?

**Review by lmdaly — COMMENTED** [review]
Looks really good! A few small change suggestions if you want to take a look

File: pkg/kubelet/cm/numamanager/numamanger.go (hunk 6 above)
Comment (nolancon) [reply pkg/kubelet/cm/numamanager/numamanger.go:84]: buildMaskHolder function added.

**Pull request merged**